shadowsocks/config.go

```diff
@@ -1,76 +1,198 @@
-/**
- * Created with IntelliJ IDEA.
- * User: clowwindy
- * Date: 12-11-2
- * Time: 上午10:31
- * To change this template use File | Settings | File Templates.
- */
 package shadowsocks
 
 import (
 	"encoding/json"
 	"fmt"
 	"io/ioutil"
-	// "log"
+	"net"
 	"os"
-	"reflect"
+	"sort"
 	"strings"
 	"time"
+
+	"go.uber.org/zap"
 )
 
+// Config is used in both ss-server & ss-local server, notice the different role in ss
+// can help you get the right config in differen situation
+// 1) ss server:
+// ServerPort and Password shoud be placed in order if multi user is enabled
+// 2) ss local server:
+// Server and ServerPort and Password shoud be care when is used in ss local server module
+//
+// NOTICE if the config file is setted, the config option will be disabled automatically
+
+// rolling index give out the index which server will return on next rolling get server
+var roundRobinIndex int
+
+// Config implement the ss config
 type Config struct {
-	Server     interface{} `json:"server"`
-	ServerPort int         `json:"server_port"`
-	LocalPort  int         `json:"local_port"`
-	Password   string      `json:"password"`
-	Method     string      `json:"method"` // encryption method
-	Auth       bool        `json:"auth"`   // one time auth
+	Server          string `json:"server_addr"`     // shadowsocks remote Server address, for multi server split them with comma
+	ServerPort      string `json:"server_port"`     // shadowsocks remote Server port, split with comma when multi user is enabled
+	Local           string `json:"local_addr"`      // shadowsocks local socks5 Server address
+	LocalPort       int    `json:"local_port"`      // shadowsocks local socks5 Server port
+	Password        string `json:"password"`        // shadowsocks remote server password, for multi server password should plase in order and split eith comma
+	Method          string `json:"method"`          // shadowsocks encryption method, split by comma if multi server is enabled
+	Timeout         int    `json:"timeout"`         // shadowsocks connection timeout
+	MultiServerMode string `json:"MultiServerMode"` // shadowsocks client multi-server access mode: fastest,round-robin,dissable
+	DNSServer       string `json:"dns_server"`      // shadowsocks remote Server DNS server option, if set to nil, the system DNS will be uesd for domain lookup
+
+	servers []string
+	methods []string // methods array, match the server port
 
 	// following options are only used by server
-	PortPassword map[string]string `json:"port_password"`
-	Timeout      int               `json:"timeout"`
+	PortPassword map[string]string `json:"port_password"` // shadowsocks mutli user password
 
 	// following options are only used by client
+	ServerPassword map[string]string `json:"server_password"` // shadowsocks local mutli server password
```

Comment:
There has a compatible problem with this config, which used to be assigned directly from the json config file, [example](https://github.com/shadowsocks/shadowsocks-go/blob/master/sample-config/client-multi-server.json)

Reply:
I split them to suit the ss-python config file style

Reply:
then the README should be correctly modified

Reply:
@dszhengyu fixed in 8796767